pkg/hubble/peer/service.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package peer
+
+import (
+	peerpb "github.com/cilium/cilium/api/v1/peer"
+	"github.com/cilium/cilium/pkg/node/manager"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// Service implements the peerpb.PeerServer gRPC service.
+// TODO: implement tests. This requires turning manager.Manager into an
+// interface so that it can be mocked for the purpose of testing.
+type Service struct {
+	stop chan struct{}
+	mgr  *manager.Manager
+}
+
+// Ensure that Service implements the peerpb.PeerServer interface.
+var _ peerpb.PeerServer = (*Service)(nil)
+
+// NewService creates a new Service.
+func NewService(mgr *manager.Manager) *Service {
+	return &Service{
+		stop: make(chan struct{}),
+		mgr:  mgr,
+	}
+}
+
+// Notify implements peerpb.Peer_PeerServer.Notify.
+func (s *Service) Notify(_ *peerpb.NotifyRequest, stream peerpb.Peer_NotifyServer) error {
+	h := newHandler()
+	defer h.Close()
+	// The node manager sends notifications upon call to Subscribe. As the
+	// handler's channel is unbuffered, make sure that the client is ready to
+	// read notifications to avoid a deadlock situation.
+	var g errgroup.Group
+	g.Go(func() error {
+		for {
+			select {
+			case <-s.stop:
+				return nil
+			case cn := <-h.C:
+				if err := stream.Send(cn); err != nil {
```

Comment:
Why isn't a `context.Context` part of the `Send` parameters?

Reply:
Because the [`ServerStream` interface](https://godoc.org/google.golang.org/grpc#ServerStream) doesn't use a `context.Context`. `SendMsg()` is used by `Send()`, as can be seen in the generated code in `peer.pb.go`:
```go
type Peer_NotifyServer interface {
	Send(*ChangeNotification) error
	grpc.ServerStream
}

type peerNotifyServer struct {
	grpc.ServerStream
}

func (x *peerNotifyServer) Send(m *ChangeNotification) error {
	return x.ServerStream.SendMsg(m)
}
```

Reply:
@gandro I have added a test that simulates a blocked send (`TestService_NotifyWithBlockedSend`) and it failed (which is good as it tests what we want).

To fix the problem and have the test pass, I've added small buffer (of size 10 by default but this is configurable through service options) that, when full, triggers the server to stop steaming to the client.
This eventually pushes the problem back to the client which will have to call `Notify()` again in such a case. I don't think it's a problem overall because upon calling `Notify()`, the client receives the list of all peers before receiving continuous updates so maintaining a local cache client-side is easy.

Please, have a look again at the `Notify()` implementation and tell me if you think this way of handling the problem is OK.

Reply:
Awesome, yes, the code itself looks good! :tada: We have to get a bit of experience with this, because I honestly have no clue how big the buffer size should be.

Reply:
> We have to get a bit of experience with this, because I honestly have no clue how big the buffer size should be.

Neither do I. I assumed that since the connection is over a local unix domain socket, a buffer of 10 should be sufficient. I also don't want to make it unnecessarily large because if the caller can't keep up then there's probably a problem elsewhere.